internal/libraries/gitutils/gitutils.go

```diff
@@ -156,3 +157,65 @@ func SortedCommitTags(repository *git.Repository) ([]*plumbing.Reference, error)
 
 	return sortedTags, nil
 }
+
+// CheckoutTag checks out the repository to the given tag.
+func CheckoutTag(repository *git.Repository, tag *plumbing.Reference) error {
+	repoTree, err := repository.Worktree()
+	if err != nil {
+		return err
+	}
+
+	// Annotated tags have their own hash, different from the commit hash, so the tag must be resolved before checkout
+	resolvedTag, err := resolveTag(tag, repository)
+	if err != nil {
+		return err
+	}
+
+	if err = repoTree.Checkout(&git.CheckoutOptions{Hash: *resolvedTag, Force: true}); err != nil {
+		return err
+	}
+
+	// Ensure the repository is checked out to a clean state.
+	for range [2]int{} {
+		clean, err := cleanRepository(repoTree)
+		if err != nil {
+			return err
+		}
+		if clean {
+			return nil
+		}
+	}
```

Comment:
It's not certain that there is any point in another try after a first unsuccessful cleaning attempt. However, go-git's checkout code does one hard reset:
https://github.com/go-git/go-git/blob/v5.4.1/worktree.go#L185
and we discovered that the repository can be left dirty after that hard reset under certain circumstances.

Running an identical redundant hard reset procedure again gets the repository to a clean state in this particular test case. 

So if it sometimes takes two resets to get it clean for no apparent reason, this indicates to me that there might be times when one more retry might be successful if the first clean fails.

This is a situation where I still have no clue what the problem is, even after having spent more time than I can afford on investigating it. None of the other people who reported the problem gave an idea of what the problem is and there is no indication of a pending fix from the maintainers. We need to get this working and we need that to happen fast. So I'm at a point where the only way forward I can see is to throw a workaround at the problem, hope it will fix it, and if not then at least don't put the corrupted release in the index.

Reply:
Understood, fine for me it solves the issue but I'd add a comment to make it clear why it's done. Even a link to this thread would be fine.

Reply:
Excellent point! What do you think about this?:
https://github.com/arduino/libraries-repository-engine/compare/2c9c7a670e8de40e8aa97753a9b9701b92c0ad2f..8d54b1aedf72f227d4a615726a8a772e7bf6d1b6

Reply:
Yup, that's cool. :+1: